src/workerd/api/streams/writable.c++

```diff
@@ -280,6 +282,137 @@ jsg::Ref<WritableStream> WritableStream::constructor(
   return kj::mv(stream);
 }
 
+namespace {
+
+// Wrapper around `WritableStreamSink` that makes it suitable for passing off to capnp RPC.
+class WritableStreamRpcAdapter final: public capnp::ExplicitEndOutputStream {
+public:
+  WritableStreamRpcAdapter(kj::Own<WritableStreamSink> inner)
+      : inner(kj::mv(inner)) {}
+  ~WritableStreamRpcAdapter() noexcept(false) {
+    weakRef->invalidate();
+    doneFulfiller->fulfill();
+  }
+
+  // Returns a promise that resolves when the stream is dropped. If the promise is canceled before
+  // that, the stream is revoked.
+  kj::Promise<void> waitForCompletionOrRevoke() {
+    auto paf = kj::newPromiseAndFulfiller<void>();
+    doneFulfiller = kj::mv(paf.fulfiller);
+
+    return paf.promise.attach([weakRef = weakRef->addRef()]() mutable {
```

Comment:
Oh shoot, using `.attach()` here actually just attaches the lambda object but it will never be called. This is a bug, I meant to use `kj::defer()`. :facepalm: 

Ben C even tried to make this an error at one point but that change never got merged...

Reply:
90847efe